types/node/util.d.ts

```diff
@@ -1105,6 +1105,197 @@ declare module 'util' {
+     *
+     *   - `values` A mapping of parsed option names with their string
+     *     or boolean values.
+     *   - `positionals` Positional arguments.
+     *   - `tokens` Detailed parse information (only if `tokens` was specified).
+     *
+     */
+    export function parseArgs<T extends ParseArgsConfig>(config: T): ParsedResults<T>;
+
+    interface ParseArgsOptionConfig {
+        type: 'string' | 'boolean';
+        short?: string;
+        multiple?: boolean;
+    }
+
+    interface ParseArgsOptionsConfig {
+        [longOption: string]: ParseArgsOptionConfig;
+    }
+
+    interface ParseArgsConfig {
+        strict?: boolean;
+        allowPositionals?: boolean;
+        tokens?: boolean;
+        options?: ParseArgsOptionsConfig;
+        args?: string[];
+    }
+
+    /*
+    IfDefaultsTrue and IfDefaultsFalse are helpers to handle default values for missing boolean properties.
+    TypeScript does not have exact types for objects: https://github.com/microsoft/TypeScript/issues/12936
+    This means it is impossible to distinguish between "field X is definitely not present" and "field X may or may not be present".
+    But we expect users to generally provide their config inline or `as const`, which means TS will always know whether a given field is present.
+    So this helper treats "not definitely present" (i.e., not `extends boolean`) as being "definitely not present", i.e. it should have its default value.
+    This is technically incorrect but is a much nicer UX for the common case.
+    The IfDefaultsTrue version is for things which default to true; the IfDefaultsFalse version is for things which default to false.
+    They differ only in the final item of the if-else chain.
+    */
+    type IfDefaultsTrue<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfTrue;
+
+    type IfDefaultsFalse<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfFalse;
+
+    type ExtractOptionValue<T extends ParseArgsConfig, O extends ParseArgsOptionConfig> = IfDefaultsTrue<
+        T['strict'],
+        O['type'] extends 'string' ? string : O['type'] extends 'boolean' ? boolean : string | boolean,
+        string | boolean
+    >;
+
+    type ParsedValues<T extends ParseArgsConfig> = (T['options'] extends ParseArgsOptionsConfig
+        ? {
+              -readonly [LongOption in keyof T['options']]: IfDefaultsFalse<
+                  T['options'][LongOption]['multiple'],
+                  undefined | Array<ExtractOptionValue<T, T['options'][LongOption]>>,
+                  undefined | ExtractOptionValue<T, T['options'][LongOption]>
+              >;
+          }
+        : {}) &
+        IfDefaultsTrue<T['strict'], unknown, { [longOption: string]: undefined | string | boolean }>;
```

Comment:
Since intersections are ordered, its generally better to put the weaker side of the intersection on the left (so the `IfDefaultsTrue<...>` constituent should come first because it defines a blanket string indexer).

Reply:
I'm happy to switch these on your say-so, but I don't know what "intersections are ordered" means.

Reply:
In TS, union types are unordered in that the order of the constituents in the union type should have no effect on assignability. Essentially, `A | B` and `B | A` are equivalent. Intersection types are ordered in that the checker makes a distinction between `A & B` and `B & A` in a few specific cases (generally for overload resolution, but there are a few others). As a result, its usually a good idea to maintain left-to-right ordering in intersections (less specific on the left, more specific on the right) to be on the safe side.

Reply:
Got it, thanks for the explanation.